### PR TITLE
Added Copy/Paste functionality to configuration profiles.

### PR DESCRIPTION
**Pull request by StephenHodgson**
Overview
---
- Added two buttons on the master configuration profile for options on how users would like to make their profiles. They can choose to copy from the default ones and customize, or create new profiles.
image
- Configuration profiles now have copy/paste functionality.
image
- Default Configuration Profiles will now show warnings with helpful hints about how to proceed.
image
- All Configuration Profiles can be copied into a completely new asset.
- Once a master configuration profile has been swapped out with a custom once, an option to replace default profiles appears.
image

Changes
---
- Fixes: #2856
- Fixed an issue where BoundarySystemIsEnabled wasn't checking if a profile was set.

### CODE REVIEW

File: Assets/MixedRealityToolkit/_Core/Inspectors/Profiles/MixedRealityConfigurationProfileInspector.cs [hunk 1]
```diff
@@ -116,7 +119,24 @@ public override void OnInspectorGUI()
...
+                EditorGUILayout.HelpBox("The Mixed Reality Toolkit's core SDK profiles can be used to get up and running quickly.\n\n" +
+                                        "You can use the default profiles provided, copy and customize the default profiles, or create your own from scratch", MessageType.Warning);
+                EditorGUILayout.BeginHorizontal();
+
+                if (GUILayout.Button("Copy & Customize"))
+                {
+                    CreateCopyProfileValues();
+                }
+
+                if (GUILayout.Button("Start from Scratch"))
```
Comment (SimonDarksideJ) [inline Assets/MixedRealityToolkit/_Core/Inspectors/Profiles/MixedRealityConfigurationProfileInspector.cs]:
```suggestion
                if (GUILayout.Button("Create a new profile"))
```
better title

Comment (StephenHodgson) [inline Assets/MixedRealityToolkit/_Core/Inspectors/Profiles/MixedRealityConfigurationProfileInspector.cs]:
```suggestion
                if (GUILayout.Button("Start from blank"))
```

File: Assets/MixedRealityToolkit/_Core/Inspectors/Profiles/MixedRealityConfigurationProfileInspector.cs [hunk 2]
```diff
@@ -116,7 +119,24 @@ public override void OnInspectorGUI()
 
             if (MixedRealityPreferences.LockProfiles && !((BaseMixedRealityProfile)target).IsCustomProfile)
             {
-                EditorGUILayout.HelpBox("The Mixed Reality Toolkit's core SDK profiles can be used to get up and running quickly.\n\nYou can use the default profiles provided or create your own in the context menu:\n'Create/Mixed Reality Toolkit/...'", MessageType.Warning);
+                EditorGUILayout.HelpBox("The Mixed Reality Toolkit's core SDK profiles can be used to get up and running quickly.\n\n" +
+                                        "You can use the default profiles provided, copy and customize the default profiles, or create your own from scratch", MessageType.Warning);
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit/_Core/Inspectors/Profiles/MixedRealityConfigurationProfileInspector.cs]:
```suggestion
                                        "You can use the default profiles provided, copy and customize the default profiles, or create your own.", MessageType.Warning);
```

**Comment by davidkline-ms** [comment]
I pulled this change and do not see the new buttons...  am I missing sonething?

**Comment by StephenHodgson** [comment]
Did you have custom profiles?

**Comment by davidkline-ms** [comment]
> Did you have custom profiles?

My steps were:
1. Pull your branch
2. Load the project
3. Click Mixed Reality Toolkit > Configure
-- no copy button --

Note: My profiles are unlocked.

**Comment by StephenHodgson** [comment]
Lock the profiles and select the MixedRealityManager

**Comment by davidkline-ms** [comment]
> Lock the profiles and select the MixedRealityManager's configuration profile.

Why do they have to be locked for this feature to work? It feels confusing. Also, shouldn't I be able to clone custom profiles?

**Comment by StephenHodgson** [comment]
>Why do they have to be locked for this feature to work?

Because it only works if they're unlocked.

**Comment by davidkline-ms** [comment]
> Because it's only meant to work if they're locked. That's by design.

I am not sure i agree with the design. i found it confusing why the behavior difference... i believe it will confuse our customers as well.

**Comment by davidkline-ms** [comment]
what is the difference between Copy and Create Copy From in the drop down menu?

**Comment by StephenHodgson** [comment]
image
This is simply a warning about the profile being one of the default ones. Users will get the choice to copy and Customize or simply create their own, or use the ones provided as is.
Every profile has this warning, but the master configuration profile is unique with these buttons because it's the only one that doesn't have a "Copy & Customize" button to press on the MixedRealityToolkit GameObject.
image

The full copy/paste exists as a context menu to the Object itself, so pressing the gear in the top right makes since cause it's the same way you'd copy and paste component values for anything else.
image

**Comment by StephenHodgson** [comment]
I'm open to updating the text to be "Clone & Customize" for the button.

and updating the context menu to read as "Create a Clone from Profile Values

**Comment by wassx** [comment]
Love this feature!

File: Assets/MixedRealityToolkit/_Core/Inspectors/MixedRealityToolkitInspector.cs [hunk 3]
```diff
@@ -53,7 +53,7 @@ public override void OnInspectorGUI()
                                                                   "Would you like to create one now?", "OK", "Later"))
                     {
                         ScriptableObject profile = CreateInstance(nameof(MixedRealityToolkitConfigurationProfile));
-                        profile.CreateAsset();
+                        profile.CreateAsset("Assets/CustomProfiles");
```
Comment (wassx) [inline Assets/MixedRealityToolkit/_Core/Inspectors/MixedRealityToolkitInspector.cs]: Minor naming thing. Maybe call it something like "Assets/MixedRealityToolkit-generated/CustomProfiles"
to be aware where it belongs to in a bigger folder structure and it's generated?

Comment (StephenHodgson) [reply Assets/MixedRealityToolkit/_Core/Inspectors/MixedRealityToolkitInspector.cs]:
```suggestion
                        profile.CreateAsset("Assets/MixedRealityToolkit-generated/CustomProfiles");
```

**Comment by keveleigh** [comment]
>No one outside of our development team will likely unlock the profiles, nor should they (unless they're modifying it for a PR against the MRTK)

We can't prevent that. Having the copy button only appear when the profile is locked is confusing, since it's the "default" profile regardless of locked state, and will be overwritten in a Toolkit update.

It's almost more important to provide information about creating a copied profile in the unlocked state, since that's the "danger" state where you might make changes that won't persist.

**Comment by davidkline-ms** [comment]
> Having the copy button only appear when the profile is locked is confusing, since it's the "default" profile regardless of locked state

This is the root of my confusion. There is no easily discoverable way to set isCustomProfile to true if the Copy button is hidden when profiles are unlocked.

I would be 100% in support of this change if the copy button existed in both the locked and unlocked states.

**Comment by StephenHodgson** [comment]
I disagree as the only state a new profile is in is `isCustomProfile==tue`

**Comment by StephenHodgson** [comment]
> I would be 100% in support of this change if the copy button existed in both the locked and unlocked states.

Technically the buttons are always there in the context menu. But sure I'll take a PR with your suggested edits.

**Comment by StephenHodgson** [comment]
@davidkline-ms added button visibility when unlocked.

File: Assets/MixedRealityToolkit/_Core/Inspectors/Profiles/MixedRealityBaseConfigurationProfileInspector.cs [hunk 4]
```diff
@@ -17,6 +22,11 @@ public abstract class MixedRealityBaseConfigurationProfileInspector : BaseMixedR
         [SerializeField]
         private Texture2D logoDarkTheme = null;
 
+        private static BaseMixedRealityProfile profile;
+        private static SerializedObject targetProfile;
+
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit/_Core/Inspectors/Profiles/MixedRealityBaseConfigurationProfileInspector.cs]:
```suggestion
```

**Comment by davidkline-ms** [comment]
> @davidkline-ms added button visibility when unlocked.

Great! i'll try it at my first opportunity,

**Comment by StephenHodgson** [comment]
@davidkline-ms going to let you test and merge this then satisfactory :)

**Comment by davidkline-ms** [comment]
Just pulled this PR and I do not see copy buttons. Tested the profiles being locked and unlocked.

**Comment by davidkline-ms** [comment]
wait... I may have opened the wrong branch... let me double-check

**Comment by davidkline-ms** [comment]
Yep. Wrong branch. Buttons appear in both states. Thanks!

**Comment by wassx** [comment]
Will this find its way into the mrtk_development branch or will the mrtk_2018beta_stabilization be the most recent branch?

**Pull request merged**